lib/IRGen/MetadataRequest.cpp

```diff
@@ -512,15 +512,28 @@ CanType IRGenModule::getRuntimeReifiedType(CanType type) {
   }));
 }
 
+static bool containsPrivateType(CanType ty) {
+  return ty.findIf([](Type t) -> bool {
+    if (auto *nominal = t->getAnyNominal()) {
+      if (nominal->getEffectiveAccess() <= AccessLevel::FilePrivate)
+        return true;
+    }
+    return false;
+  });
+}
+
 CanType IRGenModule::substOpaqueTypesWithUnderlyingTypes(CanType type) {
   // Substitute away opaque types whose underlying types we're allowed to
   // assume are constant.
   if (type->hasOpaqueArchetype()) {
     ReplaceOpaqueTypesWithUnderlyingTypes replacer(getSwiftModule(),
                                                   ResilienceExpansion::Maximal);
-    type = type.subst(replacer, replacer,
-                      SubstFlags::SubstituteOpaqueArchetypes)
-      ->getCanonicalType();
+    auto underlyingTy =
```

Comment:
Since you have an IRGenModule here, I think you should instead use `getTypeInfo(type).isABIAccessible()`

Reply:
isABIAccessible does not give the right answer for private non resilient types:

```
 /// Whether this type is known to be ABI-accessible, i.e. whether it's        
  /// actually possible to do ABI operations on it from this current SILModule. 
  /// See SILModule::isTypeABIAccessible.                                       
  ///                                                                           
  /// All fixed-size types are currently ABI-accessible, although this would    
  /// not be difficult to change (e.g. if we had an archetype size constraint   
  /// that didn't say anything about triviality).                               
  IsABIAccessible_t isABIAccessible() const {    
```

The failure occurs when we build the associated type witness. The associated type resolves to an opaque type and the underlying type contains the private type.

Reply:
Ah, good point. In that case, I think it's still worth extracting a new predicate since it might come up elsewhere.

Is 'private' access enough though? It seems if the declaration is internal, you'll run into the same problem.

Reply:
You are right internal access is problematic across modules (the problematic case can happen with non-inlineable functions from non resilient modules).

I will fix the code in the ReplaceOpaqueTypesWithUnderlyingTypes utility instead.

Reply:
New PR: https://github.com/apple/swift/pull/27666